lib/Conversion/LinalgToArith/LinalgToArith.cpp

```diff
@@ -0,0 +1,385 @@
+    //     diagonal_elements.push_back(diagonal_element)
+    //
+    // Finally, we create a new constant op with the diagonalized elements.
+
+    if (intType) {
+      SmallVector<APInt> diagonalElements;
+      diagonalElements.reserve(dim0 * dim1);
+      for (int i = 0; i < transposedDimensions[0]; ++i) {
+        for (int j = 0; j < transposedDimensions[1]; ++j) {
+          int index =
+              calculateIndexHelper(leftOperandIsSecret, dim0, dim1, i, j);
+          auto value = denseAttr.getValues<APInt>()[index];
+          diagonalElements.push_back(value);
+        }
+      }
+      auto diagonalizedType = RankedTensorType::get(transposedDimensions, type);
+      auto diagonalizedDenseElementsAttr =
+          DenseElementsAttr::get(diagonalizedType, diagonalElements);
+      newMatrix = b.create<arith::ConstantOp>(diagonalizedDenseElementsAttr);
+    } else if (floatType) {
+      SmallVector<APFloat> diagonalElements;
+      diagonalElements.reserve(dim0 * dim1);
+      for (int i = 0; i < transposedDimensions[0]; ++i) {
+        for (int j = 0; j < transposedDimensions[1]; ++j) {
+          int index =
+              calculateIndexHelper(leftOperandIsSecret, dim0, dim1, i, j);
+          auto value = denseAttr.getValues<APFloat>()[index];
+          diagonalElements.push_back(value);
+        }
+      }
+
+      auto diagonalizedType = RankedTensorType::get(transposedDimensions, type);
+      auto diagonalizedDenseElementsAttr =
+          DenseElementsAttr::get(diagonalizedType, diagonalElements);
+      newMatrix = b.create<arith::ConstantOp>(diagonalizedDenseElementsAttr);
+    }
+
+    // Now, we create code for multiplying the matrix with rotations of the
+    // vector.
+
+    // The code below emits the following code for vector-matrix
+    // multiplication (matrix-vector multiplication is similar):
+    // %sum = bias
+    // %rotated_vector = secretValues
+    // for %i = 0 to dim1 - 1:
+    //   %extractedSlice = extract_slice %newMatrix[%i, 0] [1, dim0] [1, 1]
+    //   %multiplied = %rotated_vector * %extractedSlice
+    //   %sum = %sum + %multiplied
+    //   %rotated_vector = rotate %rotated_vector, 1
+    // %lastExtracted = extract_slice %newMatrix[dim1-1, 0] [1, dim0] [1, 1]
+    // %final_sum = %sum + %lastExtracted
+    // At this point, we can rotate and sum if needed. (Squat packing is left as
+    // a TODO until we resolve the shape mismatch issue.)
+    // return %final_sum
+
+    // Build a constant index 1.
+    auto indexOne = b.create<arith::ConstantIndexOp>(1);
+
+    // Setup sizes and strides, which are parameters for the tensor_ext
+    // ExtractSliceOp.
+    SmallVector<OpFoldResult> sizes(2);
+    if (leftOperandIsSecret) {
+      sizes = {rewriter.getIndexAttr(1), rewriter.getIndexAttr(dim0)};
+    } else {
+      sizes = {rewriter.getIndexAttr(dim1), rewriter.getIndexAttr(1)};
+    }
+    SmallVector<OpFoldResult> strides(
+        {rewriter.getIndexAttr(1), rewriter.getIndexAttr(1)});
```

Comment:
```suggestion
       2, rewriter.getIndexAttr(1));
```

(i.e. (num, element))